README.md
TrackJS Browser Agent
===============

The browser agent for collecting errors from JavaScript, Network, and console
exceptions. The agent wraps the browser API to record context about your
application, network, and visitor, and include that context when unhandled
errors occur.

You'll need a TrackJS account to use this agent. If you don't have one, [why not
start a free trial today?](https://trackjs.com/signup)

## Quick Start

1. [Signup for TrackJS](https://trackjs.com/signup) and get your token.
2. Add a dependency on TrackJS with `npm install trackjs --save`
3. Install the agent in your app. You can paste the script tags into the
`<head>` of your HTML:

```html
<script src="PATH_TO_TRACKJS/t.js"></script>
<script>
  window.TrackJS && TrackJS.install({ token: 'YOUR_TOKEN_HERE' });
</script>
```

Or you can bundle it as a module into your application.

```javascript
import { TrackJS } from 'trackjs';
TrackJS.install({ token: 'YOUR_TOKEN_HERE' });
```

4. Test it out by calling `TrackJS.track('testing!')` somewhere in your
application.
5. You should see your error on [TrackJS](https://my.trackjs.com/recent) within
a few seconds.


## More Information

You can find more information about how to install and configure the agent in
the [TrackJS Documentation](https://docs.trackjs.com/). If you run into any
trouble, let us know right away at `hello@trackjs.com`



CHANGELOG.md
JavaScript Browser Agent Changelog
==================================

## [Version 3.10.4](https://cdn.trackjs.com/agent/3.10.4/t.js) (2023-12-05)

Update Typescript types for Network optional fields.

## [Version 3.10.3](https://cdn.trackjs.com/agent/3.10.3/t.js) (2023-11-13)

No Change, build system updates.

## [Version 3.10.2](https://cdn.trackjs.com/agent/3.10.2/t.js) (2023-10-23)

Features:
  - Verified with Node 18 LTS

## [Version 3.10.1](https://cdn.trackjs.com/agent/3.10.1/t.js) (2021-08-05)

Bugfixes:
  - Fix typo in readme.

## [Version 3.10.0](https://cdn.trackjs.com/agent/3.10.0/t.js) (2021-07-15)

Features:
  - Agent supports ServiceWorkers and SharedWorkers

## [Version 3.9.2](https://cdn.trackjs.com/agent/3.9.2/t.js) (2021-05-01)

Bugfixes:
  - Handle case where read-only error object passed into the global error handler.
  - Handle serializing a circular-referenced object with broken property getters.

## [Version 3.9.1](https://cdn.trackjs.com/agent/3.9.1/t.js) (2021-04-23)

Bugfixes:
  - Fixed rare case where `Illegal invocation` was thrown when calling a console function.

## [Version 3.9.0](https://cdn.trackjs.com/agent/3.9.0/t.js) (2021-03-04)

Features:
  - Added `forwardingDomain` option to allow for ad blocker bypass.

## [Version 3.8.0](https://cdn.trackjs.com/agent/3.8.0/t.js) (2021-02-17)

Features:
  - Agent supports use inside of WebWorkers. Can be imported as a module or using `importScript`.
  - Tooling upgrade to Node 14.5.4

## [Version 3.7.7](https://cdn.trackjs.com/agent/3.7.7/t.js) (2021-02-11)

Bugfixes:
  - Normalized the error messages for network failures, which sometimes do not include a `statusText` property, by removing it from the message string. Customers will no longer need duplicate ignore rules to handle both variants.
  - Expose error when user calls `XMLHttpRequest.send` before `XMLHttpRequest.open`, which was previously treated as an internal fault.
  - Update typescript definitions to include the `error` argument in the `onError` callback.

## [Version 3.7.6](https://cdn.trackjs.com/agent/3.7.6/t.js) (2020-11-30)

Bugfixes:
  - Handles internal failure where the reason for a Promise rejection is `null`.

## [Version 3.7.5](https://cdn.trackjs.com/agent/3.7.5/t.js) (2020-09-10)

Bugfixes:
  - Escapes space and tab characters in network telemetry URLs that would be automatically converted by the browser.

## [Version 3.7.4](https://cdn.trackjs.com/agent/3.7.4/t.js) (2020-05-19)

Bugfixes:
  - Fix `fetch` wrapping that caused errors during abort. ([Issue#38](https://github.com/TrackJs/trackjs-package/issues/38)).
  - Add console.warn Typescript definition. ([Issue#39](https://github.com/TrackJs/trackjs-package/issues/39)).

## [Version 3.7.3](https://cdn.trackjs.com/agent/3.7.3/t.js) (2020-02-26)

Bugfixes:
  - Allow `undefined` configuration properties to be passed without stopping.


## [Version 3.7.2](https://cdn.trackjs.com/agent/3.7.2/t.js) (2019-09-12)

Bugfixes:
  - Update `fetch` wrapping to accommodate non-standardized URL object behavior utilized by Facebook sdk.


## [Version 3.7.1](https://cdn.trackjs.com/agent/3.7.1/t.js) (2019-09-11)

Bugfixes:
  - Change `fetch` wrapping to avoid a [Microsoft Edge Bug](https://developer.microsoft.com/en-us/microsoft-edge/platform/issues/23391036/) creating a `Request` object.
  - Add `TrackJS.isInstalled()` to Typescript definition.


## [Version 3.7.0](https://cdn.trackjs.com/agent/3.7.0/t.js) (2019-08-19)

Features:
  - Capture the element inner text as part of visitor Telemetry.


## [Version 3.6.0](https://cdn.trackjs.com/agent/3.6.0/t.js) (2019-08-12)

Bugfixes:
  - Add safety checking to XHR header detection to not show console warnings.


## [Version 3.5.0](https://cdn.trackjs.com/agent/3.5.0/t.js) (2019-08-12)

Features:
  - Revert of 3.4.0 due to a CORS issue.


## [Version 3.4.0](https://cdn.trackjs.com/agent/3.4.0/t.js) (2019-08-05)

Features:
  - Check for the presence of a Node correlationId on network request for cross-referencing with the Node agent.
  - Force TrackJS agent version into dependencies, even if we were bundled.
  - Change method of Image creation to be safer.
  - Added `agentPlatform` to error capture payload.
  - Added `TrackJS.isInstalled()` for checking install state.

Bugfixes:
  - None


## [Version 3.3.0](https://cdn.trackjs.com/agent/3.3.0/t.js) (2019-03-04)

Features:
  - Changed error capture to accept a `202 Accepted` response from the API to be more accurate to behavior.
  - Added agent version to error capture API call.
  - Improved internal implementation of util.contains.

Bugfixes:
  - Handle case where undefined can be thrown from an XHR listener.


## [Version 3.2.0](https://cdn.trackjs.com/agent/3.2.0/t.js) (2019-02-22)

Features:
  - Support Symbol types in object serializer.

Bugfixes:
  - Fixed error thrown when requiring into Node. Replaced with no-op.


## [Version 3.1.2](https://cdn.trackjs.com/agent/3.1.2/t.js) (2019-01-31)

Features:
  - None

Bugfixes:
  - Fixed serialization of unthrown errors in IE11.


## [Version 3.1.1](https://cdn.trackjs.com/agent/3.1.1/t.js) (2019-01-26)

Features:
  - None

Bugfixes:
  - Changed how window and document are detected in the main closure for webpack.



## [Version 3.1.0](https://cdn.trackjs.com/agent/3.1.0/t.js) (2019-01-19)

Features:
  - Detect if we are running in a NodeJS environment and safely disable monitoring.

Bugfixes:
  - Removed warning message printed when functions are called where `enabled=false` is in the configuration.



## [Version 3.0.1](https://cdn.trackjs.com/agent/3.0.1/t.js) (2018-11-10)

Features:
  - None

Bugfixes:
  - Updated install example with a safety check.
  - Updated types definition to include the global object.


## [Version 3.0.0](https://cdn.trackjs.com/agent/3.0.0/t.js) (2018-10-29)

Features:
  - Changed the global namespace to `TrackJS`. The previous `trackJs` still exists, but is deprecated.

tracker.d.ts
// Type definitions for tracker.js
// Project: https://github.com/TrackJS/trackjs-package

/** String formatted as an ISO-8601 Date. Example 0000-00-00T00:00:00.000Z */
interface ISO8601DateString extends String {}

/**
 * Payload of an error sent to TrackJS. Useful when manipulating errors via
 * the `onError` callback.
 */
interface TrackJSPayload {

  /**
   * Stack trace at time of asynchronous callback binding.
   */
  bindStack?: string;

  /**
   * Timestamp of the asynchronous callback binding.
   */
  bindTime?: ISO8601DateString;

  /**
   * Telemetry logs gathered from the console.
   */
  console: {

    /** Timestamp the event occurred */
    timestamp: ISO8601DateString;

    /** Console severity of the event */
    severity: string;

    /** Formatted message captured */
    message: string;

  }[];

  /**
   * Context provided about the current customer (you)
   */
  customer: {

    /** Customer application id */
    application?: string;

    /** Unique Id describing the current page view */
    correlationId: string;

    /** Customer-provided visitor session ID */
    sessionId?: string;

    /** Customer token */
    token: string;

    /** Customer-provided visitor user ID */
    userId?: string;

    /** Customer-provided system version ID */
    version?: string;

  };

  /**
   * How the error was captured. Can be "window","direct","global", "ajax" or
   * "catch"
   */
  entry: string;

  /**
   * Context about the browser environment
   */
  environment: {

    /** How long the visitor has been on the page in MS */
    age: number;

    /**
     * Other discovered JavaScript libraries on the DOM. Hashmap of
     * name: version pairs.
     */
    dependencies: { [name: string]: string };

    /** browser userAgent string */
    userAgent: string;

    /** current window height */
    viewportHeight: number;

    /** current window width */
    viewportWidth: number;

  };

  /**
   * Custom environment metadata.
   */
  metadata: {

    /** metadata group name */
    key: string;

    /** metadata value */
    value: string;

  }[];

  /** Error message */
  message: string;

  /**
   * Telemetry logs gathered from the network.
   */
  network: {

    /** Timestamp the request started */
    startedOn: ISO8601DateString;

    /** Timestamp the request completed */
    completedOn: ISO8601DateString;

    /** HTTP Method used */
    method: string;

    /** URL Requested */
    url: string;

    /** HTTP Status Code */
    statusCode: number;

    /** HTTP Status Text */
    statusText: string;

  }[];

  /** location of the browser at the time of the error */
  url: string;

  /** stack trace */
  stack: string;

  /** client-reported time the error occurred */
  timestamp: ISO8601DateString;

  /**
   * Telemetry logs gathered from the visitor.
   */
  visitor: {

    /** timestamp the event occurred */
    timestamp: ISO8601DateString;

    /** visitor action taken. "input" or "click" */
    action: string;

    /** DOM element acted upon */
    element: {

      /** name of the element tag. IE "input" */
      tag: string;

      /** hashmap of element attributes */
      attributes: { [attributeName: string]: string };

      /** value of the element */
      value: {

        /** Number of characters in the value */
        length: number;

        /** Patterns describing the value. */
        pattern: string;

      };

    };

  }[];

  /** version of the tracker.js lib */
  version: string;

  /** Number of messages throttled clientside */
  throttled: number;

}

/**
 * Configuration options that can be passed to `trackJs.configure()`
 */
interface TrackJSOptions {

  /**
   * Whether duplicate errors should be suppressed before sending.
   * default true.
   */
  dedupe?: boolean;

  /**
   * Custom handler to be notified *before* an error is transmitted. Can be used
   * to modify or ignore error data.
   */
  onError?: (payload: TrackJSPayload) => boolean;

  /**
   * Custom handler for serializing non-string data in errors and telemetry
   * events.
   */
  serialize?: (what: any) => string;

  /**
   * Customer-provided identification of the visitor session. Use this to
   * correlate TrackJS Error reports with other reporting data.
   */
  sessionId?: string;

  /**
   * Customer-provided identification of the visitor. Use this to identify the
   * current user.
   */
  userId?: string;

  /**
   * Customer-provided identification of the running system. Use this to
   * identify the version of the code running. Recommend to use either a SEMVER
   * representation, or a VCS Hash Key.
   */
  version?: string;

}

/**
 * Configuration options that are initialized from `window._trackJs`
 */
interface TrackJSInitOptions extends TrackJSOptions {

  /**
   * Your account token. Get this from `https://my.trackjs.com/install`
   */
  token: string;

  /**
   * Whether the tracker script is enabled. Default true.
   */
  enabled?: boolean;

  /**
   * TrackJS Application token. Get this from `https://my.trackjs.com/Account/Applications`
   */
  application?: string;

  /**
   * Options for recording errors from native callback interfaces, such as
   * `setTimeout` and `addEventListener`.
   */
  callback?: {

    /**
     * Whether errors should be recorded when caught from callback functions.
     * default true.
     */
    enabled?: boolean;

    /**
     * Whether stack traces should be generated at the time of invocation of an
     * asynchronous action. This will produce stack traces similar to the
     * "async" traces in Chrome Developer tools.
     * There is a performance impact to enabling this. Confirm behavior in your
     * application before releasing.
     * default false.
     */
    bindStack?: boolean;

  };

  /**
   * Options for recording errors and telemetry from the console.
   */
  console?: {

    /**
     * Whether events should be recorded from the console.
     * default true.
     */
    enabled?: boolean;

    /**
     * Whether console messages should be passed through to the browser console
     * or hidden by TrackJS. Useful for removing debug messaging from production.
     * default true.
     */
    display?: boolean;

    /**
     * Whether an error should be recorded by a call to `console.error`.
     * default true.
     */
    error?: boolean;

    /**
     * Limit the console functions to be watched by whitelisting them here.
     * default ["log","debug","info","warn","error"]
     */
    watch?: string[];

  };

  /**
   * Options for recording errors and telemetry from the network.
   */
  network?: {

    /**
     * Whether events should be recorded from the network.
     * default true.
     */
    enabled?: boolean;

    /**
     * Whether an error should be recorded by XHR responses with status code
     * 400 or greater.
     * default true.
     */
    error?: boolean;

  };

  /**
   * Options for recording telemetry from the visitor.
   */
  visitor?: {

    /**
     * Whether events should be recorded from the visitor actions.
     * default true.
     */
    enabled?: boolean;

  };

  /**
   * Options for recording errors from the global window.
   */
  window?: {

    /**
     * Whether events should be recorded from globally unhandled errors.
     * default true.
     */
    enabled?: boolean;

    /**
      * Whether events should be recorded from globally unhandled promise
      * rejections, if supported. default true.
      */
    promise?: boolean;

  };

}

/**
 * The TrackJS global namespace for functions.
 */
interface TrackJSStatic {

  /**
   * Adds a new key-value pair to the metadata store. If the key already exists
   * it will be updated.
   *
   * @param {String} key
   * @param {String} value
   */
  addMetadata(key: string, value: string): void;

  /**
   * Invokes the provided function within a try/catch wrapper that forwards
   * the error to TrackJS.
   *
   * @param {Function} func The function to be invoked.
   * @param {Object} context The context to invoke the function with.
   * @param {...} Additional arguments passed to the function.
   * @return {*} Output of the function.
   */
  attempt(func: Function, context?: any, ...args: any[]): any;

  /**
   * Configures the instance of TrackJS with the provided configuration.
   *
   * @param {Object} options The Configuration object to apply
   * @returns {Boolean} True if the configuration was successful.
   */
  configure(options: TrackJSOptions): boolean;

  /**
    * Non-exposed browser console logging. Use this private console to prevent
    * messages from being exposed into the standard browser console.
    */
  console: {

    /**
      * Records context into the Telemetry log with normal severity
      *
      * @param {...} args Arguments to be serialized into the Telemetry log.
      */
    log(...args: any[]): void;

    /**
      * Records context into the Telemetry log with DEBUG severity
      *
      * @param {...} args Arguments to be serialized into the Telemetry log.
      */
    debug(...args: any[]): void;

    /**
     * Records context into the Telemetry log with INFO severity
     *
     * @param {...} args Arguments to be serialized into the Telemetry log.
     */
    info(...args: any[]): void;

    /**
     * Records context into the Telemetry log with WARN severity
     *
     * @param {...} args Arguments to be serialized into the Telemetry log.
     */
    warn(...args: any[]): void;

    /**
     * Records context into the Telemetry log with ERROR severity. If console
     * errors are enabled, which is default, this will also transmit an error.
     *
     * @param {...} args Arguments to be serialized into the Telemetry log.
     */
    error(...args: any[]): void;

  };

  /**
   * Removes a key from the metadata store, if it exists.
   *
   * @param {String} key
   */
  removeMetadata(key: string): void;

  /**
   * Directly invokes an error to be sent to TrackJS.
   *
   * @param {Error|Object|String} error The error to be tracked. If error does
   *        not have a stacktrace, will attempt to generate one.
   */
  track(error: Error | Object | String): void;

  /**
   * Returns a wrapped and watched version of the function to automatically
   * catch any errors that may arise.
   *
   * @param {Function} func The function to be watched.
   * @param {Object} context The context to invoke the function with.
   * @return {Function} Wrapped function
   */
  watch(func: Function, context?: any): Function;

  /**
   * Wrap and watch all of the functions on an object that will
   * automatically catch any errors that may arise.
   *
   * @param {Object} obj The Object containing functions to be watched
   * @return {Object} Object now containing wrapped functions.
   */
  watchAll(obj: Object): Object;

  /**
   * Running version of the tracker script
   */
  version: string;

}

/**
 * TrackJS initialization object, hanging off of the window.
 */
interface Window {
  _trackJs: TrackJSInitOptions
}

declare var trackJs: TrackJSStatic;
declare module "Tracker" {
  export = trackJs;
}


index.d.ts
// Type definitions for TrackJS 3.10.4
// Project: https://github.com/TrackJS/trackjs-package

interface TrackJSStatic {

  /**
   * Adds a new key-value pair to the metadata store. If the key already exists
   * it will be updated.
   *
   * @param {String} key
   * @param {String} value
   */
  addMetadata(key: string, value: string): void;

  /**
   * Invokes the provided function within a try/catch wrapper that forwards
   * the error to TrackJS.
   *
   * @param {Function} func The function to be invoked.
   * @param {Object} context The context to invoke the function with.
   * @param {...} Additional arguments passed to the function.
   * @return {*} Output of the function.
   */
  attempt(func: Function, context?: any, ...args: any[]): any;

  /**
   * Configures the instance of TrackJS with the provided configuration.
   *
   * @param {Object} options The Configuration object to apply
   * @returns {Boolean} True if the configuration was successful.
   */
  configure(options: TrackJSConfigureOptions): boolean;

  /**
   * Non-exposed browser console logging. Use this private console to prevent
   * messages from being exposed into the standard browser console.
   */
  console: {

    /**
     * Records context into the Telemetry log with normal severity
     *
     * @param {...} args Arguments to be serialized into the Telemetry log.
     */
    log(...args: any[]): void;

    /**
     * Records context into the Telemetry log with DEBUG severity
     *
     * @param {...} args Arguments to be serialized into the Telemetry log.
     */
    debug(...args: any[]): void;

    /**
     * Records context into the Telemetry log with INFO severity
     *
     * @param {...} args Arguments to be serialized into the Telemetry log.
     */
    info(...args: any[]): void;

    /**
     * Records context into the Telemetry log with WARN severity
     *
     * @param {...} args Arguments to be serialized into the Telemetry log.
     */
    warn(...args: any[]): void;

    /**
     * Records context into the Telemetry log with ERROR severity. If console
     * errors are enabled, which is default, this will also transmit an error.
     *
     * @param {...} args Arguments to be serialized into the Telemetry log.
     */
    error(...args: any[]): void;

  }

  /**
   * Running version of the tracker script
   */
  hash: string;

  /**
   * Installs the agent into the current browser document.
   *
   * @param options The configuration object to apply.
   */
  install(options: TrackJSInstallOptions): boolean;

  /**
   * Whether the agent has been installed into the current environment
   *
   * @returns {boolean}
   */
  isInstalled(): boolean;

  /**
   * Removes a key from the metadata store, if it exists.
   *
   * @param {String} key
   */
  removeMetadata(key: string): void;

  /**
   * Directly invokes an error to be sent to TrackJS.
   *
   * @param {Error|Object|String} error The error to be tracked. If error does
   *        not have a stacktrace, will attempt to generate one.
   */
  track(error: Error | Object | String): void;

  /**
   * Running version of the tracker script
   */
  version: string;

  /**
   * Returns a wrapped and watched version of the function to automatically
   * catch any errors that may arise.
   *
   * @param {Function} func The function to be watched.
   * @param {Object} context The context to invoke the function with.
   * @return {Function} Wrapped function
   */
  watch(func: Function, context?: any): Function;

  /**
   * Wrap and watch all of the functions on an object that will
   * automatically catch any errors that may arise.
   *
   * @param {Object} obj The Object containing functions to be watched
   * @return {Object} Object now containing wrapped functions.
   */
  watchAll(obj: Object): Object;
}

/**
 * String formatted as an ISO-8601 Date. Example 0000-00-00T00:00:00.000Z
 */
export interface ISO8601DateString extends String { }

/**
 * Payload of an error sent to TrackJS. Useful when manipulating errors via
 * the `onError` callback.
 */
export interface TrackJSErrorPayload {
  /** Stack trace at time of asynchronous callback binding. */
  bindStack?: string;
  /** Timestamp of the asynchronous callback binding. */
  bindTime?: ISO8601DateString;
  /** Browser Console Telemetry */
  console: {
    /** Timestamp the event occurred */
    timestamp: ISO8601DateString;
    /** Console severity of the event */
    severity: string;
    /** Formatted message captured */
    message: string;
  }[];
  /** Context provided about the customer session */
  customer: {
    /** Customer application id */
    application?: string;
    /** Unique Id describing the current page view */
    correlationId: string;
    /** Customer-provided visitor session ID */
    sessionId?: string;
    /** Customer token */
    token: string;
    /** Customer-provided visitor user ID */
    userId?: string;
    /** Customer-provided system version ID */
    version?: string;
  };
  /** How the error was captured. */
  entry: string;
  /** Context about the browser environment */
  environment: {
    /** How long the visitor has been on the page in MS */
    age: number;
    /** Other discovered JavaScript libraries on the DOM. */
    dependencies: { [name: string]: string };
    /** browser userAgent string */
    userAgent: string;
    /** current window height */
    viewportHeight: number;
    /** current window width */
    viewportWidth: number;
  };
  /** Custom environment metadata. */
  metadata: {
    /** metadata group name */
    key: string;
    /** metadata value */
    value: string;
  }[];
  /** Error message */
  message: string;
  /** Navigation Telemetry */
  nav: {
    /** Timestamp of the navigation event */
    "on": ISO8601DateString;
    /** Navigation method used. IE "replaceState" "setState" */
    "type": string;
    /** Previous URL */
    "from": string;
    /** New URL */
    "to": string
  }[];
  /** Network Telemetry */
  network: {
    /** Timestamp the request started */
    startedOn: ISO8601DateString;
    /** Timestamp the request completed */
    completedOn?: ISO8601DateString;
    /** HTTP Method used */
    method: string;
    /** URL Requested */
    url: string;
    /** HTTP Status Code */
    statusCode?: number;
    /** HTTP Status Text */
    statusText?: string;
    /** Mechanism of network use. IE "fetch", "xhr" */
    type: string;
  }[];
  /** location of the browser at the time of the error */
  url: string;
  /** stack trace */
  stack: string;
  /** client-reported time the error occurred */
  timestamp: ISO8601DateString;
  /** Visitor Action Telemetry */
  visitor: {
    /** timestamp the event occurred */
    timestamp: ISO8601DateString;
    /** visitor action taken. "input" or "click" */
    action: string;
    /** DOM element acted upon */
    element: {
      /** name of the element tag. IE "input" */
      tag: string;
      /** hashmap of element attributes */
      attributes: { [attributeName: string]: string };
      /** value of the element */
      value: {
        /** Number of characters in the value */
        length: number;
        /** Patterns describing the value. */
        pattern: string;
      };
    };
  }[];
  /** version of the tracker.js lib */
  version: string;
  /** Number of messages throttled clientside */
  throttled: number;
}

/**
 * Configuration options that can be passed to `trackJs.configure()`
 */
export interface TrackJSConfigureOptions {

  /**
   * Whether duplicate errors should be suppressed before sending.
   * @default true
   */
  dedupe?: boolean;

  /**
   * Whether to attempt discovery of other JavaScript libs on the page.
   * @default true
   */
  dependencies?: boolean;

  /**
   * Custom handler to be notified *before* an error is transmitted. Can be used
   * to modify or ignore error data.
   *
   * @param {TrackJSErrorPayload} payload Error payload to send to TrackJS.
   * @param {Error} error Error object that initiated the capture.
   */
  onError?: (payload: TrackJSErrorPayload, error?: Error) => boolean;

  /**
   * Custom handler for serializing non-string data in errors and telemetry
   * events.
   */
  serialize?: (what: any) => string;

  /**
   * Id of the visitor session. Use this to correlate TrackJS
   * Error reports with other reporting data.
   */
  sessionId?: string;

  /**
   * Id of the visitor. Use this to identify the current user for support.
   */
  userId?: string;

  /**
   * Id of the running application. Recommend to use either a SEMVER
   * representation, or a VCS Hash Key.
   */
  version?: string;

}

/**
 * Configuration options that are initialized from `window._trackJs`
 */
export interface TrackJSInstallOptions extends TrackJSConfigureOptions {

  /**
   * TrackJS Application token. Get this from `https://my.trackjs.com/Account/Applications`
   */
  application?: string;

  callback?: {
    /**
     * Whether errors should be recorded when caught from callback functions.
     * @default true
     */
    enabled?: boolean;
    /**
     * Whether stack traces should be generated at the time of invocation of an
     * asynchronous action. This will produce stack traces similar to the
     * "async" traces in Chrome Developer tools.
     * There is a performance impact to enabling this. Confirm behavior in your
     * application before releasing.
     * default false.
     */
    bindStack?: boolean;
  };

  console?: {
    /**
     * Whether events should be recorded from the console.
     * @default true
     */
    enabled?: boolean;
    /**
     * Whether console messages should be passed through to the browser console
     * or hidden by TrackJS. Useful for removing debug messaging from production.
     * @default true
     */
    display?: boolean;
    /**
     * Whether an error should be recorded by a call to `console.error`.
     * @default true
     */
    error?: boolean;
    /**
     * Whether a warning should be recorded by a call to `console.warn`.
     * @default false
     */
    warn?: boolean;
    /**
     * Limit the console functions to be watched by whitelisting them here.
     * @default ["log","debug","info","warn","error"]
     */
    watch?: string[];
  };

  /**
   * Whether the tracker script is enabled.
   * @default true
   */
  enabled?: boolean;

  /**
   * Domain where errors and usage data to be sent in order to bypass ad blocker
   * restrictions to `trackjs.com`. IE "errors.example.com".
   */
  forwardingDomain?: string;

  network?: {
    /**
     * Whether events should be recorded from the network.
     * @default true
     */
    enabled?: boolean;
    /**
     * Whether an error should be recorded by XHR responses with status code
     * 400 or greater.
     * @default true
     */
    error?: boolean;
  };

  /**
   * Your account token. Get this from `https://my.trackjs.com/install`
   */
  token: string;

  visitor?: {
    /**
     * Whether events should be recorded from the visitor actions.
     * @default true
     */
    enabled?: boolean;
  };

  window?: {
    /**
     * Whether events should be recorded from globally unhandled errors.
     * @default true
     */
    enabled?: boolean;
    /**
      * Whether events should be recorded from globally unhandled promise
      * rejections, if supported.
      * @default true
      */
    promise?: boolean;
  };

}

declare global {
  var TrackJS: TrackJSStatic | undefined;
  interface Window {
    TrackJS: TrackJSStatic | undefined;
  }
}

declare var TrackJS: TrackJSStatic
export { TrackJS }
